backend/tests/services/test_local_device_command_service.py

```
@@ -324,6 +324,277 @@ def test_local_device_command_registry_default_includes_diagnostic_commands():
+
+def test_workspace_tree_script_rejects_non_workspace_cwd(tmp_path, monkeypatch, capsys):
+    """workspace_tree should reject arbitrary directories outside workspace roots."""
+    import json
+
+    from app.services.device.command_registry import WORKSPACE_TREE_SCRIPT
+
+    monkeypatch.delenv("WEGENT_WORKSPACE_ROOTS", raising=False)
+    monkeypatch.chdir(tmp_path)
+
+    try:
+        exec(WORKSPACE_TREE_SCRIPT, {"__name__": "__main__"})
```

Ruff (0.15.15)

[error] Use of `exec` detected

(S102)